### PR TITLE
Add go tool tests and CircleCI config to vet

### LINKED ISSUE #100: Auto build and push all used docker images.
Currently we build and push images on every code change. 
Lets use circle or Travis to automate this.

### LINKED ISSUE #152: Fix and add test for go vet errors.
This would be a good test to add as a first with travis.

first suggested in: https://github.com/prometheus/prombench/pull/150

this might be helpful for other tests.
https://github.com/diazjf/fakeclient

### PR DESCRIPTION
**Pull request by CrazyByDefault**
Closes #152 , continuation of PR #161 
closes #161
closes #100

I've added a simple CircleCI config that runs make vet, and have updated the Makefile.common to the latest one used on prometheus master

Requesting review @krasi-georgiev :)

### CODE REVIEW

**Comment by CrazyByDefault** [comment]
I've split the CircleCI config into two jobs - _test_ and _build_

_test_ will run `make vet` and `make build`

_build_ will run `docker build` and push it to the appropriate dockerhub registry. Note that a CircleCI context will need to be added named `docker`, with the vars `DOCKER_LOGIN` and `DOCKER_PWD` to be used for docker login so that once merged, CircleCI can push to prombench registry

_build_ will only run on versioned tags on any branch (eg. v2.0) Tag validated by regex `/^v[0-9]+(\.[0-9]+)*$/`

Closes #100 if this behaviour is ok

**Comment by simonpasquier** [comment]
Thanks for the contribution, I'll have a look at it next week!

**Comment by krasi-georgiev** [comment]
to close #100 we need to build and push all images from 
https://hub.docker.com/u/prombench
but lets do this in a separate PR anyway.

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+          command: |
+            docker build -f Dockerfile -t $DOCKER_LOGIN/$CIRCLE_PROJECT_REPONAME:$CIRCLE_TAG .
+            echo $DOCKER_PWD | docker login -u $DOCKER_LOGIN --password-stdin
+              docker push $DOCKER_LOGIN/$CIRCLE_PROJECT_REPONAME:$CIRCLE_TAG
+
+workflows:
+  version: 2
+  build_workflow:
+    jobs:
+      - test
```
Comment (krasi-georgiev) [inline .circleci/config.yml]: what is this job for?

File: .promu.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+go:
+    version: 1.11.0
```
Comment (simonpasquier) [inline .promu.yml:2]: you should only specify the major and minor versions here (eg 1.11, 1.12).

File: Makefile [hunk 3]
```diff
@@ -1,10 +1,17 @@
...
 
+include Makefile.common
+
+# Prombench Commands
 ifeq ($(AUTH_FILE),)
-AUTH_FILE = "/etc/serviceaccount/service-account.json"
+	AUTH_FILE = "/etc/serviceaccount/service-account.json"
 endif
 
+.PHONY: deploy clean
```
Comment (simonpasquier) [inline Makefile:13]: (nit) I tend to have the `.PHONY` line next to the target to which it applies. In other words, `.PHONY: clean` should go right before the `clean: ...` definition.

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    working_directory: /go/src/github.com/{{ORG_NAME}}/{{REPO_NAME}}
+    steps:
+      - checkout
+      - run: make build 
+      - setup_remote_docker:
+          docker_layer_caching: true
+      - run:
+          name: build docker image
+          command: |
+            docker build -f Dockerfile -t $DOCKER_LOGIN/$CIRCLE_PROJECT_REPONAME:$CIRCLE_TAG .
```
Comment (simonpasquier) [inline .circleci/config.yml:29]: It would make sense to reuse the existing recipes from `Makefile.common`.

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+      - run: make build
+
+  build:
+    docker:
+      - image: circleci/golang:1.11
+
+    working_directory: /go/src/github.com/{{ORG_NAME}}/{{REPO_NAME}}
+    steps:
+      - checkout
+      - run: make build
```
Comment (simonpasquier) [inline .circleci/config.yml:23]: Usually we decouple the job building the binary from the job building the container image. And the latter reuses the output from the former.

File: .circleci/config.yml [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+version: 2
+jobs:
+  test:
+    docker:
+      - image: circleci/golang:1.11
+
+    working_directory: /go/src/github.com/{{ORG_NAME}}/{{REPO_NAME}}
+    steps:
+      - checkout
+      - run: make vet
```
Comment (simonpasquier) [inline .circleci/config.yml:13]: I would add `make unused`. `go.sum` needs to be committed too.

**Comment by krasi-georgiev** [comment]
@CrazyByDefault  don't forget that tomorrow the 9th is the last day for the GSOC proposal submissions. Feel free  to ping me on IRC with yours so I can give some feedback before submitting.

**Comment by krasi-georgiev** [comment]
@CrazyByDefault are you going to continue working on this?

**Pull request closed without merging**